### ISSUE TITLE
A new layer type to mask areas outside region of interest

### ISSUE THREAD

**Issue by planemad** [mapbox/mapbox-gl-js#6267]
This is a feature request. The implementation details are unknown.

## Motivation

As a map designer, a common cartographic technique for me to focus the user attention to an area of interest is to mute/hide details of the areas outside the region of interest. 

To accomplish such an effect with the existing GL style-spec, one needs to create a polygon covering the whole world with a hole of the desired region. This is a cumbersome and tedious data creation process for a map designer.

## Design Alternatives

- A new `inverted-fill` layer type which will apply a fill style to all areas outside of a polygon similar to [qgis](http://oslandia.com/en/2014/06/14/masking-features-in-qgis-2-4/)
- New `inverted-fill-x` properties for all existing fill styles
- Do nothing is not an option

## Design

Feel that maybe a new `inverted-fill` layer type might be the simplest with respect to keeping the style-spec clean. 

### Mock-Up

What will this design look like to developers?

```
map.addLayer({
	'id': 'area of interest',
	'type': 'inverted-fill',
	'source': 'aoi',
	'layout': {},
	'paint': {
		'fill-pattern': 'hatch',
		'fill-opacity': 0.8
	}
});
```

What will this design look like to end users?
![](https://i.stack.imgur.com/hjfVT.jpg)
_[Image source: inverted fill in qgis](https://gis.stackexchange.com/questions/153754/inverted-shading-with-polygons-in-qgis-2-8)_

### Concepts

`inverted-fill` is the new concept: Fill the inverted area of the polygon. This feature is called "inverted polygon" in [qgis](http://oslandia.com/en/2014/06/14/masking-features-in-qgis-2-4/).

**Comment by anandthakker** [comment]
Another alternative would be to add a `fill-region: 'inside' | 'outside'` property to the existing fill layer.

**Comment by andrewharvey** [comment]
This feature would make it easier for the map designer, avoiding the need to use turf.mask. See prior discussion for this feature at #993 and #6257.

**Comment by ryanbaumann** [comment]
While masks work great for maps without symbol layers, one challenge are labels near the borders of the mask.  This proposal is likely out-of-scope for this particular bug, but it's important to consider when creating masks.

![](https://cl.ly/0h273e0X2e3X/download/Image%202018-03-03%20at%208.25.00%20AM.png)

**Comment by markusjohnsson** [comment]
We need this feature in our application and I am willing to work on it to make it happen. 

What we need is to invert the 'water' layer - we need a 'land' layer to place above a custom layer where we render wave height and ocean currents. And because it is such a large dataset for the water layer, we would really like to not have to download, invert offline and upload it again.

(Or, we would need a way to mask our custom layer with the water layer, but I think an inverted `fill` is more usable.)

I have implemented a proof-of-concept (/hack) of @mourner 's suggestion from https://github.com/mapbox/mapbox-gl-js/issues/993#issuecomment-74130793 to invert geojson on-the-fly inside mapbox, but did not manage to do the same with vector tiles.. Would that be feasible? 

I would like some guidance so that I don't develop something that cannot be merged.

**Comment by markusjohnsson** [comment]
I've worked on this a bit more, taking another approach than the one mentioned above. Instead I'm using the stencil buffer to mask the polygon and then fill the tile.

![image](https://user-images.githubusercontent.com/557685/49152109-16a9fc80-f312-11e8-962a-e2e4e5f10676.png)

Still lots of work to be done, but at least this looks like it will work.

**Comment by waissbluth** [comment]
Is this feature planned?

**Comment by markusjohnsson** [comment]
@waissbluth I've received no indication that this would be accepted if I opened a pull request. I still need this for a customer project, but it is on low priority right now.

**Comment by unitof** [comment]
@markusjohnsson Do you have a public fork you'd be willing to share? Looks like a wondrous start—is the "inverse" fill styled the same way I would style a `fill`-type layer?

**Comment by 1ec5** [comment]
This is orthogonal to the request for masking, but in case anyone is coming here looking for a way to effectively dim the background layer outside an area of interest without blotting it out entirely, I shared a simple approach towards the end of [this workshop](https://github.com/mapbox/workshops/blob/gh-pages/sotmus-2018-heatmaps/README.md#back-to-studio) at State of the Map U.S. 2018.

![detroit](https://user-images.githubusercontent.com/1231218/56827384-8a91f780-6813-11e9-8e6b-54236db96496.png)

**Comment by fvonk** [comment]
Is there any feature like this (mask areas outside polygon) in Mapbox iOS SDK ?
Appreciate for the answer!

**Comment by andrewharvey** [comment]
The new `within` expression https://blog.mapbox.com/introducing-gl-js-v1-9-0-and-the-within-expression-6268f6c32be3 may help with some of the use cases here. It will let you apply a different style to points and lines outside a polygon.

**Comment by fvonk** [comment]
> The new `within` expression https://blog.mapbox.com/introducing-gl-js-v1-9-0-and-the-within-expression-6268f6c32be3 may help with some of the use cases here. It will let you apply a different style to points and lines outside a polygon.

Sorry, but I mean in Mapbox iOS SDK, not in mapbox-gl-sj.

**Comment by andrewharvey** [comment]
> Sorry, but I mean in Mapbox iOS SDK, not in mapbox-gl-sj.

I was commenting in general about this thread, not in response to your question.

**Comment by 1ec5** [comment]
@fvonk, as of mapbox/mapbox-gl-native-ios#184 and v5.8.0, you can use a `SELF IN %@` expression, which is equivalent to the `within` expression seen here. However, it only works on whole features in shape or vector tile sources; it won’t cut up a single feature that happens to be partially inside your region of interest.

**Comment by Stophface** [comment]
You can use turf's mask function for that
https://www.npmjs.com/package/@turf/mask
Simply create a "cover polygon" with the extend of your (Map)View (or the whole world) and use your Polygon. Turf even creates the "cover polygon" for you, if you do not provide one.
https://github.com/Turfjs/turf/blob/master/packages/turf-mask/index.js#L58

**Comment by s11richard** [comment]
Has there been any update to the status of this feature request?

We are currently using the solution that leverages turf-mask to invert the geometries. But in our case, it is too cumbersome to render on the client, so we are processing them as tilesets.

In order to avoid the high cost of processing 10m resolution tiles that span the entire planet, we are forced to use a maxzoom value of 5. This results in a 300m resolution, which is not ideal.

This feature seems like the only solution to our problem.

**Comment by boredland** [comment]
It's a shame that `within` can only be applied to point and line geometries, not everything else.

**Comment by chriszrc** [comment]
@boredland yeah, and that lines won't get clipped to the bounds either, they're either all in or out :(

**Comment by david-morris** [comment]
This issue is making the use of vector tiles much more difficult.  I'll have to make two sets for now, and needing to do so programattically rules out Mapbox's web UI for doing so.

**Comment by firelite** [comment]
I'm looking for this exact feature! I want to be able to grey out anything that's not within certain areas, say around certain cities and counties where our services aren't available yet.
